src/Catalyst.Core.Modules.Keystore/KeyChain.cs

```diff
@@ -0,0 +1,497 @@
+        {
+            var key = await Store.TryGetAsync(name, cancel).ConfigureAwait(false);
+            if (key == null)
+                throw new KeyNotFoundException($"The key '{name}' does not exist.");
+            AsymmetricKeyParameter kp = null;
+            UseEncryptedKey(key, pkey => { kp = pkey; });
+            return kp;
+        }
+
+        void UseEncryptedKey(EncryptedKey key, Action<AsymmetricKeyParameter> action)
+        {
+            using (var sr = new StringReader(key.Pem))
+            using (var pf = new PasswordFinder
+            {
+                Password = _dek
+            })
+            {
+                var reader = new PemReader(sr, pf);
+                var privateKey = (AsymmetricKeyParameter) reader.ReadObject();
+                action(privateKey);
+            }
+        }
+
+        async Task<IKey> AddPrivateKeyAsync(string name, AsymmetricCipherKeyPair keyPair, CancellationToken cancel)
+        {
+            // Create the key ID
+            var keyId = CreateKeyId(keyPair.Public);
+
+            // Create the PKCS #8 container for the key
+            string pem;
+            using (var sw = new StringWriter())
+            {
+                var pkcs8 = new Pkcs8Generator(keyPair.Private, Pkcs8Generator.PbeSha1_3DES)
+                {
+                    Password = _dek
+                };
+                var pw = new PemWriter(sw);
+                pw.WriteObject(pkcs8);
+                await pw.Writer.FlushAsync();
+                pem = sw.ToString();
+            }
+
+            // Store the key in the repository.
+            var key = new EncryptedKey
+            {
+                Id = keyId.ToBase58(),
+                Name = name,
+                Pem = pem
+            };
+            await Store.PutAsync(name, key).ConfigureAwait(false);
+            log.DebugFormat("Added key '{0}' with ID {1}", name, keyId);
+
+            return new KeyInfo
+            {
+                Id = key.Id, Name = key.Name
+            };
+        }
+
+        /// <summary>
+        ///   Create a key ID for the key.
+        /// </summary>
+        /// <param name="key"></param>
+        /// <returns></returns>
+        /// <remarks>
+        ///   The key id is the SHA-256 multihash of its public key. The public key is 
+        ///   a protobuf encoding containing a type and 
+        ///   the DER encoding of the PKCS SubjectPublicKeyInfo.
+        /// </remarks>
+        MultiHash CreateKeyId(AsymmetricKeyParameter key)
+        {
+            var spki = SubjectPublicKeyInfoFactory
+               .CreateSubjectPublicKeyInfo(key)
+               .GetDerEncoded();
+
+            // Add protobuf cruft.
+            var publicKey = new Lib.Cryptography.Proto.PublicKey
+            {
+                Data = spki
+            };
+            if (key is RsaKeyParameters)
+                publicKey.Type = Lib.Cryptography.Proto.KeyType.RSA;
+            else if (key is ECPublicKeyParameters)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)